### PR TITLE
Vendor `intl` NumberFormat

### LINKED ISSUE #1241: Vendor `intl`
### Description

Reason https://github.com/getsentry/sentry-dart/pull/1202
https://github.com/dart-lang/pub/issues/3747
Ps: Check the license permission.

### PR DESCRIPTION
**Pull request by denrase**
#skip_changelog

## :scroll: Description

_#skip-changelog_

Closes #1241

## :bulb: Motivation and Context

Implement specific method to format sample rate.

## :green_heart: How did you test it?

Unit tests with data created form `NumberFormat` output.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

We could come up with more input/output string combinations.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,10 @@
 
 - View hierarchy reads size from RenderBox only ([#1258](https://github.com/getsentry/sentry-dart/pull/1258))
 
+### Chores
+
+- Replace intl NumberFormat ([#1269](https://github.com/getsentry/sentry-dart/pull/1269))
```
Comment (marandaneto) [inline CHANGELOG.md]: I'd say we can skip the changelog here, its just refactoring, non user facing.

File: dart/lib/src/utils/sample_rate_format.dart [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import 'package:meta/meta.dart';
+
+@internal
+class SampleRateFormat {
+  static String format(double sampleRate) {
+    final fixed = sampleRate.toStringAsFixed(16);
+    return fixed.replaceAll(RegExp(r"([.]*0+)(?!.*\d)"), "");
```
Comment (marandaneto) [inline dart/lib/src/utils/sample_rate_format.dart]: Are we using the same Regex from `NumberFormat`?

Comment (denrase) [reply dart/lib/src/utils/sample_rate_format.dart]: NumberFormat does a bunch of things. This regex just removes trailing zeros from the string.

**Comment by denrase** [comment]
@marandaneto I randomly generated numbers to see if the formatting is the same. I have encountered an oddity from number formatter, where the 16th place after the decimal is rounded up/down. See below for a comparison. To me it looks that `NumberFormat` is doing the wring thing, as it is changing the input where it's not necessary. Sometimes it's rounding values >= 5 up, and sometimes down, which might come down to double imprecision in the algorithms.

| Input                     | NumberFormat              | SampleRateFormat          |
| ------------------------- | ------------------------- | ------------------------- |
| 0.980500684268974<b>9</b> | 0.980500684268974<b>8</b> | 0.980500684268974<b>9</b> |
| 0.977855746200201<b>5</b> | 0.977855746200201<b>6</b> | 0.977855746200201<b>5</b> |
| 0.975242934280097<b>5</b> | 0.975242934280097<b>4</b> | 0.975242934280097<b>5</b> |
| 0.959001772506301<b>3</b> | 0.959001772506301<b>2</b> | 0.959001772506301<b>3</b> |

File: dart/lib/src/utils/sample_rate_format.dart [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  static String format(double sampleRate) {
+    final rounded = dp(sampleRate, 16);
+    final fixed = rounded.toStringAsFixed(16);
+    return fixed.replaceAll(RegExp(r"([.]*0+)(?!.*\d)"), "");
+  }
+
+  static double dp(double val, int places){
+    num mod = pow(10.0, places);
+    return ((val * mod).round().toDouble() / mod);
+  }
```
Comment (marandaneto) [inline dart/lib/src/utils/sample_rate_format.dart]: It does a `pow`, `round`, etc, by reading https://api.dart.dev/stable/2.19.1/dart-math/pow.html I am not sure that the output is the same as:

```
final formatter = NumberFormat('#.################');
return formatter.format(sampleRate);
```

File: dart/lib/src/utils/sample_rate_format.dart (hunk 2 above)
Comment (marandaneto) [reply dart/lib/src/utils/sample_rate_format.dart]: But the `dp` does a bunch of things and I am unsure if this has the same result, can you reason around the `pow` and `round / mod`?

**Comment by denrase** [comment]
Opted to vendor all the needed files with just minimal changes in the imports and folder structure. It's a lot of code just to get the exact same result as before on the 16th place after the comma. I'm not sure if this is worth it TBH.

**Comment by marandaneto** [comment]
> Opted to vendor all the needed files with just minimal changes in the imports and folder structure. It's a lot of code just to get the exact same result as before on the 16th place after the comma. I'm not sure if this is worth it TBH.

I get the trade off and is indeed a lot of code, but if using the previous implementation, we need to be sure that we are not running into rounding issues, the output should be the same as the `intl` parse, can we do that?

**Comment by denrase** [comment]
@marandaneto Only thing we could try is to generate a larger number of random sample values in the tests, including `intl` in the test target and compare the outputs, using a range that we ate comfortable with as rounding error. 🤔

**Comment by marandaneto** [comment]
> @marandaneto Only thing we could try is to generate a larger number of random sample values in the tests, including `intl` in the test target and compare the outputs, using a range that we ate comfortable with as rounding error. 🤔

mmm, the question is if we can avoid the routing error, unless `intl` does have the very same problem?

**Comment by denrase** [comment]
@marandaneto I stripped out all unnecessary code we need for sample rate formatting. In the tests we are formatting and comparing 1000 values randomly with `intl` (dev dependency).

File: dart/test/utils/sample_rate_format_test.dart [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      Tuple(0.19191919191, '0.19191919191'),
+      Tuple(0.191919191919, '0.191919191919'),
+      Tuple(0.1919191919191, '0.1919191919191'),
+      Tuple(0.19191919191919, '0.19191919191919'),
+      Tuple(0.191919191919191, '0.191919191919191'),
+      Tuple(0.1919191919191919, '0.1919191919191919'),
+      Tuple(0.19191919191919199, '0.191919191919192'),
+    ];
+
+    for (var inputAndOutput in inputsAndOutputs) {
```
Comment (marandaneto) [inline dart/test/utils/sample_rate_format_test.dart]:
```suggestion
    for (final inputAndOutput in inputsAndOutputs) {
```

File: dart/test/utils/sample_rate_format_test.dart [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      expect(SampleRateFormat().format(inputAndOutput.i), inputAndOutput.o);
+    }
+  });
+
+  test('formats same as intl', () {
+    for (var i = 0; i < 1000; i++) {
+      var doubleValue = Random().nextDouble();
```
Comment (marandaneto) [inline dart/test/utils/sample_rate_format_test.dart]:
```suggestion
      final doubleValue = Random().nextDouble();
```

File: dart/lib/src/utils/sample_rate_format.dart [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+  /// Helper to get the floor of a number which might not be num. This should
+  /// only ever be called with an argument which is positive, or whose abs()
+  ///  is negative. The second case is the maximum negative value on a
+  ///  fixed-length integer. Since they are integers, they are also their own
+  ///  floor.
+  dynamic _floor(dynamic number) {
+    if (number.isNegative && !number.abs().isNegative) {
+      throw ArgumentError(
+          'Internal error: expected positive number, got $number');
```
Comment (marandaneto) [inline dart/lib/src/utils/sample_rate_format.dart:189]: Either we do a try-catch when calling SampleRateFormat (and log it) or we swallow this exception returning null.

File: dart/lib/src/utils/sample_rate_format.dart [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+  }
+
+  SampleRateFormat._(this._symbols, this._zeroOffset)
+      : _minimumIntegerDigits = 1,
+        _maximumFractionDigits = 16,
+        _minimumFractionDigits = 0;
+
+  /// Format the sample rate
+  String format(dynamic sampleRate) {
+    if (_isNaN(sampleRate)) return _symbols.NAN;
```
Comment (marandaneto) [inline dart/lib/src/utils/sample_rate_format.dart]: Should we return 0 or null then? the `nan` is invalid for Sentry.

File: dart/test/utils/sample_rate_format_test.dart [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      expect(
+        NumberFormat('#.################').format(doubleValue),
+        SampleRateFormat().format(doubleValue),
```
Comment (marandaneto) [inline dart/test/utils/sample_rate_format_test.dart]: I wonder if we need that now, comparing both to guarantee that we've gotten the right algorithm makes sense (during development), but comparing was more a case of black box testing, do we need to run this every time if we know that it works as expected now? your call to keep it or remove it.

Comment (denrase) [reply dart/test/utils/sample_rate_format_test.dart]: Think we are fine removing it.

File: dart/lib/src/utils/sample_rate_format.dart (hunk 6 above)
Comment (denrase) [reply dart/lib/src/utils/sample_rate_format.dart:189]: Opted to swallow and return '0' for the sample rate, as i didn't want to make the API nullable. I don't think we will ever experience this code path, as we are only allowing values between 0 and 1.

File: dart/test/utils/sample_rate_format_test.dart [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+  test('input smaller 0 is capped', () {
+    expect(SampleRateFormat().format(-1), '0');
+  });
+
+  test('input larger 1 is capped', () {
+    expect(SampleRateFormat().format(1.1), '1');
+  });
+
+  test('call with naN returns 0', () {
```
Comment (marandaneto) [inline dart/test/utils/sample_rate_format_test.dart]:
```suggestion
  test('call with NaN returns 0', () {
```

**Review by marandaneto — APPROVED** [review]
Thanks @denrase

**Comment by marandaneto** [comment]
@denrase 

> Expected: '0.1919191919191919'
>     Actual: '0.191919191919192'
>      Which: is different.
>             Expected: ... 191919191919
>               Actual: ... 19191919192
>                                     ^
>              Differ at offset 16

File: dart/lib/src/utils/sample_rate_format.dart [hunk 10]
```diff
@@ -0,0 +1,292 @@
+import 'dart:math';
+
+import 'package:meta/meta.dart';
+
+/// Implementation details from `intl` package
+/// https://pub.dev/packages/intl
```
Comment (marandaneto) [inline dart/lib/src/utils/sample_rate_format.dart]: @chadwhitacre I'd like to double check if this is fine, License https://github.com/dart-lang/intl/blob/master/LICENSE `BSD 3`
Should we copy the license text on top of this file?

**Review by marandaneto — COMMENTED** [review]
Lets wait for https://github.com/getsentry/sentry-dart/pull/1269/files#r1142973732

**Comment by marandaneto** [comment]
> @denrase
> 
> > Expected: '0.1919191919191919'
> > Actual: '0.191919191919192'
> > Which: is different.
> > Expected: ... 191919191919
> > Actual: ... 19191919192
> > ^
> > Differ at offset 16

@denrase is this fixed? I saw some commits but not sure.

**Comment by denrase** [comment]
@marandaneto I have some weird behaviour. Since the code is from `intl`, i included it again (in tests only) to see if it also behaves differently on ubuntu and windows. CI was green for all tests then. So either the test with the rounding error is flaky or something else is going on.

**Comment by denrase** [comment]
@marandaneto I was mistaken, the test with `intl` failed with the same result. So the behaviour is the same as before and also with the vendored code. There are slightly different outcomes on different platforms. The result below is from using `intl`:

Bildschirm­foto 2023-03-21 um 14 26 21

**Comment by marandaneto** [comment]
> @marandaneto I was mistaken, the test with `intl` failed with the same result. So the behaviour is the same as before and also with the vendored code. There are slightly different outcomes on different platforms. The result below is from using `intl`:
> 
> Bildschirm­foto 2023-03-21 um 14 26 21

That's fine then, its similar to how it was before, let's compare the output of the `intl` with our code result instead of hardcoded values, this should fix the test because they should have the same result.

File: dart/lib/src/utils/sample_rate_format.dart (hunk 10 above)
Comment (chadwhitacre) [reply dart/lib/src/utils/sample_rate_format.dart]: Yes, we actually just clarified [the policy around this🔒](https://www.notion.so/sentry/Open-Source-Legal-Policy-ac4885d265cb4d7898a01c060b061e42?pvs=4#e1e3f2040b1b4f69bc888d1e3d6b5d8c) (in https://github.com/getsentry/team-ospo/issues/120):

1. Separate file
2. Link to where we got it (attribution)
3. Include license text in file header

Comment (marandaneto) [reply dart/lib/src/utils/sample_rate_format.dart]: @denrase lets do this then, make CI happy and merge it.

File: dart/pubspec.yaml [hunk 11]
```diff
@@ -25,3 +24,4 @@ dev_dependencies:
   yaml: ^3.1.0 # needed for version match (code and pubspec)
   collection: ^1.16.0
   coverage: ^1.3.0
+  intl: '>=0.17.0 <1.0.0'
```
Comment (marandaneto) [inline dart/pubspec.yaml]:
```suggestion
  intl: '>=0.17.0 <1.0.0'

```

**Comment by denrase** [comment]
@marandaneto I'm not sure why this would fail here. AFAIK the code changes are not related to this test.

Bildschirm­foto 2023-03-27 um 11 25 26

**Comment by marandaneto** [comment]
> @marandaneto I'm not sure why this would fail here. AFAIK the code changes are not related to this test.
> 
> Bildschirm­foto 2023-03-27 um 11 25 26

I already merged main into this branch, this is due to a new `dio` version, I already merged a fix for that.

**Pull request merged**